File: apps/android/app/src/main/java/pro/wewed/app/services/CapabilityPolicy.kt
package pro.wewed.app.services

import pro.wewed.app.models.AppRole
import pro.wewed.app.models.Capability
import pro.wewed.app.models.Capability.*

/**
 * Least-privilege role policy. Mirrored exactly by the iOS CapabilityPolicy;
 * the cross-platform tests assert the same matrix on both.
 */
object CapabilityPolicy {
    private val publicWeddingInfo = setOf(VIEW_WEDDING_SUMMARY, VIEW_PROGRAMME)

    private val weddingManagement = publicWeddingInfo + setOf(
        VIEW_PLANNING_DASHBOARD,
        VIEW_TASKS,
        MANAGE_TASKS,
        VIEW_BUDGET,
        VIEW_GUEST_ROSTER,
        VIEW_CONTRIBUTIONS,
        VIEW_CONTRIBUTOR_IDENTITY,
        VIEW_ALL_VENDOR_ENGAGEMENTS,
        VIEW_VENDOR_PRESENCE,
        VIEW_SEATING,
        VIEW_DOCUMENTS,
        EDIT_WEDDING_DETAILS,
        PREVIEW_GUEST_PASSES,
        VIEW_ADMISSION_SUMMARY,
        SCAN_ADMISSION,
        LOOKUP_ADMISSION
    )

    private val gateOperations = setOf(SCAN_ADMISSION, LOOKUP_ADMISSION, VIEW_ADMISSION_SUMMARY)

    fun capabilities(role: AppRole): Set<Capability> = when (role) {
        AppRole.COUPLE -> weddingManagement
        AppRole.PLANNER -> weddingManagement + setOf(PLANNER_WORKSPACE, PLANNER_ACTIONS, POST_ANNOUNCEMENT)
        AppRole.COORDINATOR -> publicWeddingInfo + gateOperations + setOf(VIEW_TASKS, VIEW_VENDOR_PRESENCE, POST_ANNOUNCEMENT)
        AppRole.VENDOR -> publicWeddingInfo + setOf(VIEW_OWN_VENDOR_ENGAGEMENT, UPDATE_OWN_VENDOR_PRESENCE)
        AppRole.USHER -> publicWeddingInfo + gateOperations
        AppRole.GUEST -> publicWeddingInfo + setOf(VIEW_OWN_INVITATION, RESPOND_OWN_RSVP, VIEW_OWN_PASS)
        AppRole.ADMIN -> setOf(VIEW_WEDDING_SUMMARY, ADMIN_SUPPORT)
    }

    fun allows(role: AppRole, capability: Capability): Boolean = capability in capabilities(role)
}

File: apps/ios/Wewed/Services/CapabilityPolicy.swift
import Foundation

/// Least-privilege role policy. Mirrors Android `CapabilityPolicy` exactly;
/// the cross-platform tests assert the same matrix on both.
public enum CapabilityPolicy {
    private static let publicWeddingInfo: Set<Capability> = [.viewWeddingSummary, .viewProgramme]

    private static let weddingManagement: Set<Capability> = publicWeddingInfo.union([
        .viewPlanningDashboard,
        .viewTasks,
        .manageTasks,
        .viewBudget,
        .viewGuestRoster,
        .viewContributions,
        .viewContributorIdentity,
        .viewAllVendorEngagements,
        .viewVendorPresence,
        .viewSeating,
        .viewDocuments,
        .editWeddingDetails,
        .previewGuestPasses,
        .viewAdmissionSummary,
        .scanAdmission,
        .lookupAdmission
    ])

    private static let gateOperations: Set<Capability> = [.scanAdmission, .lookupAdmission, .viewAdmissionSummary]

    public static func capabilities(for role: AppRole) -> Set<Capability> {
        switch role {
        case .couple:
            return weddingManagement
        case .planner:
            return weddingManagement.union([.plannerWorkspace, .plannerActions, .postAnnouncement])
        case .coordinator:
            return publicWeddingInfo.union(gateOperations).union([.viewTasks, .viewVendorPresence, .postAnnouncement])
        case .vendor:
            return publicWeddingInfo.union([.viewOwnVendorEngagement, .updateOwnVendorPresence])
        case .usher:
            return publicWeddingInfo.union(gateOperations)
        case .guest:
            return publicWeddingInfo.union([.viewOwnInvitation, .respondOwnRsvp, .viewOwnPass])
        case .admin:
            return [.viewWeddingSummary, .adminSupport]
        }
    }

    public static func allows(_ role: AppRole, _ capability: Capability) -> Bool {
        capabilities(for: role).contains(capability)
    }
}
